Analysis/IsolatedLeptonFinder/src/IsolatedLeptonFinderProcessor.cc

```diff
@@ -280,38 +324,212 @@ void IsolatedLeptonFinderProcessor::processEvent( LCEvent * evt ) {
 		}
 	}
 
-	// PFO loop
+
+	// Undressed leptons
+	std::vector<int> goodLeptonIndices;
 	int npfo = _pfoCol->getNumberOfElements();
 	for (int i = 0; i < npfo; i++ ) {
 		ReconstructedParticle* pfo = dynamic_cast<ReconstructedParticle*>( _pfoCol->getElementAt(i) );
 
-		if ( IsIsolatedLepton( pfo ) ) 
+		if ( !IsGoodLepton( pfo )){
+			otPFOsRemovedIsoLepCol->addElement( pfo );
+			continue;
+		}
+
+		if  ( IsIsolatedLepton( pfo, false ) ){
+			streamlog_out(DEBUG) << "ISOLATION undressed "<<pfo->getType()<<": SUCCESS !" << std::endl;
 			otIsoLepCol->addElement( pfo );
-		else 
+		}
+		else{
+			streamlog_out(DEBUG) << "ISOLATION undressed "<<pfo->getType()<<": FAILED" << std::endl;
 			otPFOsRemovedIsoLepCol->addElement( pfo );
+		}
+
+		// remember lepton indices for dressing
+		if (IsLepton( pfo )) goodLeptonIndices.push_back(i);
 	}
 
-	streamlog_out(DEBUG) << "   processing event: " << evt->getEventNumber() 
-		<< "   in run:  " << evt->getRunNumber() 
-		<< std::endl ;
 
-	// Add PFOs to new collection
-	evt->addCollection( otPFOsRemovedIsoLepCol, _outputPFOsRemovedIsoLepCollection.c_str() );
-	evt->addCollection( otIsoLepCol, _outputIsoLepCollection.c_str() );
-}
+	// Dressed leptons
+	// order by energy
+	for (unsigned int i = 1; i < goodLeptonIndices.size(); ++i)
+	{
+		if (isMoreEnergetic(i, i-1)) {
+			std::swap(goodLeptonIndices.at(i), goodLeptonIndices.at(i-1));
+			i = 1;
+		}
+	}
+	// dress them
+	for (unsigned int i = 0; i < goodLeptonIndices.size(); ++i)
+	{
+		ReconstructedParticle* pfo_tmp = dynamic_cast<ReconstructedParticle*>( _pfoCol->getElementAt(goodLeptonIndices.at(i) ));
```

Comment:
if you don't check for `nullptr` you can use `static_cast`

Reply:
Done